lib/SILOptimizer/Utils/Generics.cpp

```diff
@@ -174,6 +285,190 @@ createSpecializedType(CanSILFunctionType SubstFTy, SILModule &M) const {
+    Builder.addRequirement(Req, Source);
+  }
+
+  Builder.finalize(SourceLoc(), OrigGenSig->getGenericParams());
+  auto *GenericSig = Builder.getGenericSignature();
+  // Remember the new generic environment.
+  auto *GenericEnv = GenericSig->createGenericEnvironment(*M.getSwiftModule());
+
+  return std::make_pair(GenericEnv, GenericSig);
+}
+
+/// Perform some sanity checks for the requirements
+static void
+checkSpecializationRequirements(ArrayRef<Requirement> Requirements) {
+  for (auto &Req : Requirements) {
+    if (Req.getKind() == RequirementKind::SameType) {
+      auto FirstType = Req.getFirstType();
+      auto SecondType = Req.getSecondType();
+      assert(FirstType && SecondType);
+
+      bool isFirstTypeNonConcrete =
+          FirstType->hasArchetype() || FirstType->hasTypeParameter();
+      bool isSecondTypeNonConcrete =
+          SecondType->hasArchetype() || SecondType->hasTypeParameter();
+      // Only one of the types should be concrete.
+      assert((isFirstTypeNonConcrete ^ isSecondTypeNonConcrete) &&
+             "Only concrete type same-type requirements are supported by "
+             "generic specialization");
+      continue;
+    }
+
+    if (Req.getKind() == RequirementKind::Layout) {
+      continue;
+    }
+
+    llvm_unreachable("Unknown type of requirement in generic specialization");
+  }
+}
+
+ReabstractionInfo::ReabstractionInfo(SILFunction *OrigF,
+                                     ArrayRef<Requirement> Requirements) {
+  if (!OrigF->shouldOptimize()) {
+    DEBUG(llvm::dbgs() << "    Cannot specialize function " << OrigF->getName()
+                       << " marked to be excluded from optimizations.\n");
+    return;
+  }
+
+  // Perform some sanity checks for the requirements
+  checkSpecializationRequirements(Requirements);
+
+  SpecializedGenericEnv = nullptr;
+
+  OriginalF = OrigF;
+  SILModule &M = OrigF->getModule();
+  ModuleDecl *SM = M.getSwiftModule();
+  auto &Ctx = M.getASTContext();
+
+  auto OrigGenericSig = OrigF->getLoweredFunctionType()->getGenericSignature();
+  auto OrigGenericEnv = OrigF->getGenericEnvironment();
+  SpecializedGenericSig = OrigGenericSig;
+  auto ForwardingSubs = OrigGenericEnv->getForwardingSubstitutions();
+  auto ForwardingInterfaceSubsMap =
+      OrigGenericSig->getSubstitutionMap(ForwardingSubs);
+  // Map archetypes of the original function to the contextual types
+  // of the specialized function.
+  SubstitutionMap ClonerArchetypeToConcreteMap;
+  SubstitutionMap CallerArchetypeToConcreteMap;
+  SubstitutionMap CallerArchetypeToInterfaceMap;
+
+  for (auto &Req : Requirements) {
+    if (Req.getKind() == RequirementKind::SameType) {
+      auto CallerArchetype = dyn_cast<SubstitutableType>(
+          OrigGenericEnv->mapTypeIntoContext(Req.getFirstType())
+              ->getCanonicalType());
+      // Remember that a given generic parameter is mapped
+      // to a concrete type.
+      CallerArchetypeToInterfaceMap.addSubstitution(
```

Comment:
addSubstitution() is going to be private. I just refactored a dozen usages of it, please don't add new ones. Substitution maps should be constructed by using GenericSignature::getSubstitutionMap() or GenericEnvironment::getSubstitutionMap()